bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/InterleavedStorageRegenerateIndexOp.java

```diff
@@ -0,0 +1,199 @@
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ *
+ */
+package org.apache.bookkeeper.bookie;
+
+
+import io.netty.buffer.ByteBuf;
+
+import java.io.IOException;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicLong;
+
+import org.apache.bookkeeper.bookie.EntryLogger.EntryLogScanner;
+import org.apache.bookkeeper.common.util.Watcher;
+import org.apache.bookkeeper.conf.ServerConfiguration;
+import org.apache.bookkeeper.stats.NullStatsLogger;
+import org.apache.bookkeeper.util.DiskChecker;
+import org.apache.bookkeeper.util.SnapshotMap;
+import org.apache.commons.lang.time.DurationFormatUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Scan all entries in the entry log and rebuild the index file for one ledger.
+ */
+public class InterleavedStorageRegenerateIndexOp {
+    private static final Logger LOG = LoggerFactory.getLogger(InterleavedStorageRegenerateIndexOp.class);
+
+    private final ServerConfiguration conf;
+    private final long ledgerId;
+
+    public InterleavedStorageRegenerateIndexOp(ServerConfiguration conf, long ledgerId) {
+        this.conf = conf;
+        this.ledgerId = ledgerId;
+    }
+
+    public void initiate(boolean dryRun) throws IOException {
+        LOG.info("Starting index rebuilding");
+
+        DiskChecker diskChecker = Bookie.createDiskChecker(conf);
+        LedgerDirsManager ledgerDirsManager = Bookie.createLedgerDirsManager(
+                conf, diskChecker, NullStatsLogger.INSTANCE);
+        LedgerDirsManager indexDirsManager = Bookie.createIndexDirsManager(
+                conf, diskChecker,  NullStatsLogger.INSTANCE, ledgerDirsManager);
+        EntryLogger entryLogger = new EntryLogger(conf, ledgerDirsManager);
+        final LedgerCache ledgerCache;
+        if (dryRun) {
+            ledgerCache = new DryRunLedgerCache();
+        } else {
+            ledgerCache = new LedgerCacheImpl(conf, new SnapshotMap<Long, Boolean>(),
+                                              indexDirsManager, NullStatsLogger.INSTANCE);
+        }
+
+        Set<Long> entryLogs = entryLogger.getEntryLogsSet();
+
+        int totalEntryLogs = entryLogs.size();
+        int completedEntryLogs = 0;
+        long startTime = System.nanoTime();
+
+        LOG.info("Scanning {} entry logs", totalEntryLogs);
+
+        AtomicLong firstEntry = new AtomicLong(Long.MAX_VALUE);
```

Comment:
nit: since there is no concurrent requirements here, `firstEntry`, `lastEntry` and `numEntries` are just containers for primitive values. you can consider using `MutableLong` in this case. just fyi.

Reply:
MutableLong from commons? I prefer to use JDK builtins if they do the same thing.

Reply:
yeah. jdk builtins are fine